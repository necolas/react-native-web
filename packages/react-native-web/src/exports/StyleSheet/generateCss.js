/**
 * Copyright (c) 2016-present, Nicolas Gallagher.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

import hyphenateStyleName from 'hyphenate-style-name';
import mapKeyValue from '../../modules/mapKeyValue';
import normalizeValue from './normalizeValue';
import prefixStyles from '../../modules/prefixStyles';

const createDeclarationString = (prop, val) => {
  const name = hyphenateStyleName(prop);
  const value = normalizeValue(prop, val);
  if (Array.isArray(val)) {
    return val.map(v => `${name}:${v}`).join(';');
  }
  return `${name}:${value}`;
};

/**
 * Generates valid CSS rule body from a JS object
 *
 * generateCss({ width: 20, color: 'blue' });
 * // => 'color:blue;width:20px'
 */
const generateCss = style =>
  mapKeyValue(prefixStyles(style), createDeclarationString)
    .sort()
    .join(';');

export default generateCss;
