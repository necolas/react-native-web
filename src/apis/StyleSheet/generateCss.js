import hyphenate from './hyphenate';
import mapKeyValue from '../../modules/mapKeyValue';
import normalizeValue from './normalizeValue';
import prefixAll from 'inline-style-prefixer/static';

const createDeclarationString = (prop, val) => {
  const name = hyphenate(prop);
  const value = normalizeValue(prop, val);
  if (Array.isArray(val)) {
    return val.map((v) => `${name}:${v}`).join(';');
  }
  return `${name}:${value}`;
};

/**
 * Generates valid CSS rule body from a JS object
 *
 * generateCss({ width: 20, color: 'blue' });
 * // => 'color:blue;width:20px'
 */
const generateCss = (style) => mapKeyValue(prefixAll(style), createDeclarationString).sort().join(';');

module.exports = generateCss;
