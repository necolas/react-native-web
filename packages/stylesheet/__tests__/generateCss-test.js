/* eslint-env jasmine, jest */

import generateCss from '../generateCss';

describe('apis/StyleSheet/generateCss', () => {
  test('generates correct css', () => {
    const style = {
      boxShadow: '1px 1px 1px 1px #000',
      borderWidthLeft: 2,
      borderWidthRight: 3,
      position: 'absolute',
      transitionDuration: '0.1s'
    };
    expect(generateCss(style)).toMatchSnapshot();
  });
});
