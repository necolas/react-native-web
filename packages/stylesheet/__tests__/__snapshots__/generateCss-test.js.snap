// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apis/StyleSheet/generateCss generates correct css 1`] = `"-webkit-transition-duration:0.1s;border-width-left:2px;border-width-right:3px;box-shadow:1px 1px 1px 1px #000;position:absolute;transition-duration:0.1s"`;
